File: mac/Sources/SidekitCore/DataMigration.swift
import Foundation

/// One-time, best-effort relocation of the app's on-disk data from its legacy "SpeakType"
/// locations to the current "Sidekit" ones, run once at launch (see `AppPaths`).
///
/// Pure given its inputs — every URL and the `FileManager` are injected — so it's unit-tested
/// against temp directories. Each move happens only when the source exists and the destination
/// does **not**, so it never clobbers current data and is idempotent across launches. Failures
/// are swallowed (a rename must never block startup); the returned labels let the caller log
/// what moved.
public enum DataMigration {
    /// Relocate the legacy app-support directory and log file into place if needed.
    /// - Returns: labels of what moved (`"support"`, `"log"`); empty when nothing moved.
    @discardableResult
    public static func migrate(legacySupport: URL, currentSupport: URL,
                               legacyLog: URL, currentLog: URL,
                               fileManager fm: FileManager = .default) -> [String] {
        var moved: [String] = []
        if relocate(legacySupport, to: currentSupport, using: fm) { moved.append("support") }
        if relocate(legacyLog, to: currentLog, using: fm) { moved.append("log") }
        return moved
    }

    /// Move `from`→`to` exactly when `from` exists and `to` does not. Best-effort: any failure
    /// (permissions, a race) is swallowed and reported as "not moved".
    private static func relocate(_ from: URL, to: URL, using fm: FileManager) -> Bool {
        guard fm.fileExists(atPath: from.path), !fm.fileExists(atPath: to.path) else { return false }
        do {
            try fm.createDirectory(at: to.deletingLastPathComponent(), withIntermediateDirectories: true)
            try fm.moveItem(at: from, to: to)
            return true
        } catch {
            return false
        }
    }
}

File: windows/Sidekit.Core/DataMigration.cs
namespace Sidekit.Core;

/// <summary>
/// One-time relocation of per-user data from the former "SpeakType" folders to the current
/// app-name folders, run once at startup. Pure given its inputs (the roots are injected), so it is
/// unit-tested against temp directories. Each move happens only when the source exists and the
/// destination does not — it never clobbers current data and is idempotent across launches.
/// Best-effort: a failure is swallowed so a rename can never block launch; the returned labels say
/// what moved.
/// </summary>
public static class DataMigration
{
    private const string LegacyName = "SpeakType";

    /// <summary>
    /// Move <c>{root}/{legacyName}</c> → <c>{root}/{currentName}</c> under both the roaming and local
    /// roots, each only when the source exists and the destination does not.
    /// </summary>
    /// <returns>Labels of what moved (<c>"roaming"</c>, <c>"local"</c>); empty when nothing moved.</returns>
    public static IReadOnlyList<string> Migrate(string roamingRoot, string localRoot,
                                                string legacyName, string currentName)
    {
        var moved = new List<string>();
        if (Relocate(Path.Combine(roamingRoot, legacyName), Path.Combine(roamingRoot, currentName)))
        {
            moved.Add("roaming");
        }

        if (Relocate(Path.Combine(localRoot, legacyName), Path.Combine(localRoot, currentName)))
        {
            moved.Add("local");
        }

        return moved;
    }

    /// <summary>
    /// Relocate the real per-user roots (<c>%AppData%</c> and <c>%LocalAppData%</c>) from the legacy
    /// "SpeakType" name to <paramref name="currentName"/>. Call once at startup, before any
    /// settings/model/log path is read.
    /// </summary>
    public static void MigrateLegacyAppData(string currentName)
    {
        var roaming = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        var local = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        Migrate(roaming, local, LegacyName, currentName);
    }

    private static bool Relocate(string from, string to)
    {
        if (!Directory.Exists(from) || Directory.Exists(to))
        {
            return false;
        }

        try
        {
            Directory.GetParent(to)?.Create();
            Directory.Move(from, to);
            return true;
        }
        catch (IOException)
        {
            return false;
        }
        catch (UnauthorizedAccessException)
        {
            return false;
        }
    }
}
